### PR TITLE
[wip] Components package

### PR DESCRIPTION
**Pull request by severo**
Missing:
- [x] use the worker and the source maps from @hyparam/components in the app build
- [x] remove the vite boilerplate (demo)
- [x] fix the linting errors
- [x] incorporate the fixes from the HF space

Then, follow-up work:
- [ ] remove `TODO(SL)` and create issues instead
- [ ] are we OK with inlining the workers?
- [ ] publish the source files (TSX) as well as the bundle
- [ ] remove hardcoded routes like `/files?key=` or `/api/store/get?key=`
- [ ] see how to handle the CSS: which part goes to the components? which part goes to the app? will we use CSS modules (one per component)?
- [ ] format the code in a coherent way (prettier?)
- [ ] setup the CI to lint/build/deploy the lib + update the app
- [ ] setup eslint for the lib
- [ ] create a demo for the @hyparam/components lib
- [ ] incorporate new (HF-related) features from the space + authentication (headers)
- [ ] give instructions in the README of packages/components to use the library
- [ ] publish the lib
- [ ] use the lib in the space, in the hyparquet demo, in the hightable demo, and other apps
- [ ] move the app to a package in the npm workspace, instead of keeping it at the root? (reference: https://github.com/NiGhTTraX/ts-monorepo, for example)
- [ ] use vite for the app too, instead of rollup? (with https://www.npmjs.com/package/vite-node?)
- [ ] find a better name than `parsedKey` and find the correct abstraction (https://github.com/hyparam/hyperparam-cli/pull/14#discussion_r1830296648)

### CODE REVIEW

File: packages/components/package.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    "demo": "vite",
+    "lint": "eslint .",
+    "preview": "vite preview",
+    "prepublishOnly": "npm run build",
+    "test": "vitest run"
+  },
+  "peerDependencies": {
+    "hightable": "0.6.3",
+    "hyparquet": "1.5.0",
+    "hyparquet-compressors": "0.1.4",
```
Comment (platypii) [inline packages/components/package.json:43]: These should probably be `dependencies` not `peerDependencies`? When bundled I would think we want to include hightable, etc. Whereas react is likely already included by any project using these components?

File: packages/components/src/workers/parquetWorkerClient.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { asyncBufferFromUrl, cachedAsyncBuffer, AsyncBuffer, ParquetReadOptions, FileMetaData } from 'hyparquet'
+import ParquetWorker from './parquetWorker?worker'
```
Comment (platypii) [inline packages/components/src/workers/parquetWorkerClient.ts:2]: haven't seen this syntax before. what does ?worker do?

File: packages/components/src/workers/parquetWorkerClient.ts [hunk 3]
```diff
@@ -24,7 +25,7 @@ const pending = new Map<number, QueryAgent>()
 
 function getWorker() {
   if (!worker) {
-    worker = new Worker(new URL('worker.min.js', import.meta.url))
+    worker = new ParquetWorker() //new URL('./parquetWorker', import.meta.url), { type: 'module' })
```
Comment (platypii) [inline packages/components/src/workers/parquetWorkerClient.ts:28]: nit: remove commented out code

Aside: we probably can't assume the location of the worker bundle, since projects depending on it might have a different path structure. Not sure exactly how but we probably need it to be configurable.

File: packages/components/tsconfig.app.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
+  "compilerOptions": {
+    "tsBuildInfoFile": "./node_modules/.tmp/tsconfig.app.tsbuildinfo",
+    "target": "ES2020",
+    "useDefineForClassFields": true,
+    "lib": ["ES2020", "DOM", "DOM.Iterable"],
+    "module": "ESNext",
+    "skipLibCheck": true,
+
+    /* Bundler mode */
```
Comment (platypii) [inline packages/components/tsconfig.app.json:10]: not legal json

File: rollup.config.js [hunk 5]
```diff
@@ -34,18 +34,18 @@ export default [
...
+  //     format: 'umd',
+  //     sourcemap: true,
+  //   },
+  //   plugins: [
+  //     commonjs(),
+  //     resolve({ browser: true }),
+  //     terser(),
+  //     typescript(),
+  //   ],
+  // },
```
Comment (platypii) [inline rollup.config.js:50]: Maybe a different PR, but if we're using vite for components, maybe should use vite instead of rollup for cli too? (remove rollup)

File: src/streamConverters.js [hunk 6]
```diff
@@ -36,18 +36,18 @@ export function readStreamToReadableStream(fsStream) {
...
+// export async function readableStreamToArrayBuffer(input) {
+//   const reader = input.getReader()
+//   const /** @type {Uint8Array[]} */ chunks = []
+//   while (true) {
+//     const { done, value } = await reader.read()
+//     if (done) break
+//     chunks.push(value)
+//   }
+//   return new Blob(chunks).arrayBuffer()
+// }
```
Comment (platypii) [inline src/streamConverters.js:53]: delete unused code

File: packages/components/package.json (hunk 1 above)
Comment (severo) [reply packages/components/package.json]: done

File: packages/components/src/workers/parquetWorkerClient.ts (hunk 2 above)
Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts:2]: It's detailed here: https://vite.dev/guide/features.html#import-with-query-suffixes. It does not seem to be the recommended way though.

Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: fixed.

File: packages/components/src/workers/parquetWorkerClient.ts (hunk 3 above)
Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: I understand we have two options:
1. give explanations to users to include the worker in their bundler (copy the worker files, that are bundled in the lib's dist/assets/ directory, to the same directory in their app)
2. compile the worker as a base64 string.

I'll review both of them.

Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: Related discussion: https://github.com/vitejs/vite/issues/7352#issuecomment-1312236119. It's not clear what is the way to go.

Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: https://github.com/domoran/vite_worker_lib_example is interesting:

- the lib build creates the worker file in assets/, as we do
- the consumer builder (vite) is able to copy the asset by default

If we use vite for the CLI, it should be working without much config. Though, we might have to write something in the README for users that will not rely on vite for their app...

<details><summary>See the build details (bundle structure and HTML contents)</summary>
<pre><code>
➜  vite_worker_lib_example git:(master) ls
consumer  lib  README.md   
➜  vite_worker_lib_example git:(master) cd lib  
➜  lib git:(master) ✗ npm run build                                                         
                                                                                            
> lib@0.0.0 build
> tsc && vite build
                                              
vite v5.2.8 building for production...
✓ 2 modules transformed.                                                                    
dist/assets/webworker-C7uLr2oP.js  0.25 kB   
dist/lib.js                        0.46 kB │ gzip: 0.27 kB
dist/lib.umd.cjs  0.94 kB │ gzip: 0.46 kB
✓ built in 76ms
➜  lib git:(master) ✗ cd ../consumer 
➜  consumer git:(master) ✗ npm run build

> consumer@0.0.0 build
> tsc && vite build

vite v5.2.8 building for production...
✓ 5 modules transformed.
dist/assets/webworker-C7uLr2oP.js  0.25 kB
dist/index.html                    0.38 kB │ gzip: 0.27 kB
dist/assets/index-B1ZSCV_w.js      1.67 kB │ gzip: 0.69 kB
✓ built in 110ms
➜  consumer git:(master) ✗ cat dist/index.html 
&#60;!doctype html&#62;
&#60;html lang="en"&#62;
  &#60;head&#62;
    &#60;meta charset="UTF-8" /&#62;
    &#60;link rel="icon" type="image/svg+xml" href="/vite.svg" /&#62;
    &#60;meta name="viewport" content="width=device-width, initial-scale=1.0" /&#62;
    &#60;title&#62;Vite + TS&#60;/title&#62;
    &#60;script type="module" crossorigin src="/assets/index-B1ZSCV_w.js"&#62;&#60;/script&#62;
  &#60;/head&#62;
  &#60;body&#62;
    &#60;div id="app"&#62;&#60;/div&#62;
  &#60;/body&#62;
&#60;/html&#62;
</code></pre>
</details>

Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: Well, I ended inlining the worker (hence also the hyparquet and hyparquet-compressor libs) in the build lib js... The size is now:
```
dist/main.es.js  234.34 kB │ gzip: 123.03 kB │ map: 119.65 kB
dist/main.umd.js  223.48 kB │ gzip: 121.78 kB │ map: 116.58 kB
```

If it's an issue, we can instead switch to the other solution by reverting https://github.com/hyparam/hyperparam-cli/pull/14/commits/a755dfd9ee45f93e678af870796b9e40791bf6d3. The client app will have to take care of copying the build worker file to its assets (might be automatic if using Vitge, see above comment)

File: packages/components/tsconfig.app.json (hunk 4 above)
Comment (severo) [reply packages/components/tsconfig.app.json]: fixed

File: packages/components/src/workers/parquetWorkerClient.ts (hunk 2 above)
Comment (severo) [reply packages/components/src/workers/parquetWorkerClient.ts]: see https://github.com/hyparam/hyperparam-cli/pull/14#discussion_r1829193962, where I use it again.

File: packages/components/src/components/viewers/ImageView.tsx [hunk 7]
```diff
@@ -8,54 +11,64 @@ enum LoadingState {
 }
 
 interface ViewerProps {
-  file: string
-  setError: (error: Error) => void
+  parsedKey: UrlKey | FileKey
```
Comment (platypii) [inline packages/components/src/components/viewers/ImageView.tsx:14]: not sure that `parsedKey` is the name I would go for here. Maybe `content`? In the other repo I moved toward ViewerProps accepting:

```
interface ViewerProps {
  content: AsyncBufferFrom
  ...
}
export type AsyncBufferFrom = FileContent<void> | AsyncBufferFromFile
```

So basically solving the same problem but slightly different names. Ultimately what's important is that it is an object that defines how to construct an AsyncBuffer object. It needs to be a serializable object so that we can send it to the webworker. And the two cases are either: from a url or from a dropped file.

In the case of a UrlKey: it should also include a byteLength property. Why? Otherwise we might need to make a HEAD request every time we want to construct an AsyncBuffer.

This is ok for now for merging the PR. We can evolve this over time. But I think both repos have not found exactly the right abstraction here.

**Review by platypii — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -17,6 +17,7 @@ jobs:
       - uses: actions/checkout@v4
       - run: npm i
       - run: tsc
+        working-directory: ./packages/components
```
[the remark this answers is not in the document]
Comment (severo) [reply .github/workflows/ci.yml:20]: The app's code does not use typescript (apart from the jsdoc comments), so I was unsure what to do. Should we even remove tsconfig.json from the root?

File: packages/components/eslint.config.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    files: ['**/*.{ts,tsx}'],
+    languageOptions: {
+      ecmaVersion: 2020,
+      globals: globals.browser,
+    },
+    plugins: {
+      'react-hooks': reactHooks,
+      'react-refresh': reactRefresh,
+    },
+    rules: {
```
[the remark this answers is not in the document]
Comment (severo) [reply packages/components/eslint.config.js:20]: done. It's so much better now :) :) :) I should have started with that.

File: packages/components/src/main.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+export * from './components/index.ts'
```
[the remark this answers is not in the document]
Comment (severo) [reply packages/components/src/main.ts]: hmmm, in fact, I forgot to export the lib/ and workers/ directories. It depends on what we want to expose. I'll add them, and we can change as you mentioned if you prefer.

[the remark this answers is not in the document]
Comment (severo) [reply packages/components/src/main.ts]: done

File: packages/components/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+{
+  "files": [],
+  "references": [
+    { "path": "./tsconfig.app.json" },
```
[the remark this answers is not in the document]
Comment (severo) [reply packages/components/tsconfig.json:4]: no, it's the default conf by vite, but you're right, it's noise

File: rollup.config.js [hunk 12]
```diff
@@ -34,18 +34,18 @@ export default [
...
-      sourcemap: true,
-    },
-    plugins: [
-      commonjs(),
-      resolve({ browser: true }),
-      terser(),
-      typescript(),
-    ],
-  },
+  // {
```
[the remark this answers is not in the document]
Comment (severo) [reply rollup.config.js]: done

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (severo) [reply .github/workflows/ci.yml]: OK. I removed tsconfig.json. Let's see if we need it again later. Also: maybe we want to move the root app to a package too, like in https://github.com/NiGhTTraX/ts-monorepo for example

File: rollup.config.js (hunk 5 above)
Comment (severo) [reply rollup.config.js]: added to the list of tasks. Maybe with https://www.npmjs.com/package/vite-node

File: packages/components/src/components/viewers/ImageView.tsx (hunk 7 above)
Comment (severo) [reply packages/components/src/components/viewers/ImageView.tsx:14]: exactly! I couldn't find a good name, and `key` is a reserved word in React so we cannot use it to pass the data as `key={key}`.
Note that my intent is also to do the URL/path parsing once, and avoid doing it again (will be maybe more important when we will include the parsing of HF URLs, since the different parts of the URL have different meanings, see https://github.com/severo/hyparam--space/blob/main/src/lib/huggingface.ts).

Maybe the abstraction is somewhat like a filesystem (as in https://filesystem-spec.readthedocs.io): being able to handle and navigate through files stored locally, remotely on a standard HTTP hosting, on S3, on HF, etc. using the same code.

File: packages/components/README.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
+# React + TypeScript + Vite
```
[the remark this answers is not in the document]
Comment (severo) [reply packages/components/README.md]: I'll do it later today...

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -11,13 +10,6 @@ jobs:
       - run: npm i
       - run: npm run lint
 
-  typecheck:
```
Comment (platypii) [inline .github/workflows/ci.yml:14]: Why not typecheck the root project too? The files are all .js but it's still nice to check for bugs and use jsdoc types.

File: packages/components/src/components/Layout.tsx [hunk 15]
```diff
@@ -1,10 +1,11 @@
-import React, { ReactNode, useEffect } from 'react'
+import { ReactNode, useEffect } from 'react'
+import { cn } from '../lib/utils.ts'
 
 interface LayoutProps {
   children: ReactNode
   className?: string
   progress?: number
-  error?: Error
+  error?: unknown
```
Comment (platypii) [inline packages/components/src/components/Layout.tsx:8]: I understand why you did this. But it feels less type-safe?

File: packages/components/src/components/Layout.tsx [hunk 16]
```diff
@@ -33,11 +34,11 @@ export default function Layout({ children, className, progress, error, title }:
       <div className={cn('content', className)}>
         {children}
       </div>
-      <div className={cn('error-bar', error && 'show-error')}>{errorMessage}</div>
+      <div className={cn('error-bar', error !== undefined && 'show-error')}>{errorMessage}</div>
     </div>
     {progress !== undefined && progress < 1 &&
       <div className={'progress-bar'} role='progressbar'>
-        <div style={{ width: `${100 * progress}%` }} />
+        <div style={{ width: `${(100 * progress).toString()}%` }} />
```
Comment (platypii) [inline packages/components/src/components/Layout.tsx:41]: Doesn't string interpolation `${...}` implicitly call .toString() anyway? Why the change?

File: packages/components/src/lib/tableProvider.ts [hunk 17]
```diff
@@ -9,7 +11,9 @@ export function parquetDataFrame(from: AsyncBufferFrom, metadata: FileMetaData):
   const { children } = parquetSchema(metadata)
   const header = children.map(child => child.element.name)
   const sortCache = new Map<string, Promise<number[]>>()
-  const data = new Array(Number(metadata.num_rows))
+  const data = new Array<ResolvableRow>(Number(metadata.num_rows))
+  // TODO(SL) ^ isn't this too big?
+  // TODO(SL) ^ the type is currently a lie, because all rows are undefined for now
```
Comment (platypii) [inline packages/components/src/lib/tableProvider.ts:16]: In practice I haven't found this to be a problem. Worst case is (roughly) `num_rows * pointer_size` but most ML datasets are 100s of thousands to low million rows per parquet file. So only a few mb of array. Also I'm not sure but the js engine might lazy init the array. It feels fast and haven't seen it out-of-memory error here.

Fair point on the type is a lie.

**Review by platypii — APPROVED** [review]
Looks good. Only minor comments. Also there's a test failing due to multiple matches of `/` instead of just one. not sure why that changed in the dom?

Approved conditional on tests passing. :+1:

File: .github/workflows/ci.yml (hunk 14 above)
Comment (severo) [reply .github/workflows/ci.yml:14]: I had an issue: `tsc` in the CI could not recognize an import from @hyparam/components, while it works locally - as I was not sure it was important, and as it was hard to reproduce, I just removed the typescript check. I'll try to make it work.

File: packages/components/src/components/Layout.tsx (hunk 16 above)
Comment (severo) [reply packages/components/src/components/Layout.tsx]: it comes from a rule in the strict mode I added in the last commits: https://typescript-eslint.io/rules/restrict-template-expressions/. I also find it annoying. The argument is that it's too implicit. I just disabled that particular rule.

File: packages/components/src/components/Layout.tsx (hunk 15 above)
Comment (severo) [reply packages/components/src/components/Layout.tsx]: fixed

**Comment by severo** [comment]
> Looks good. Only minor comments. Also there's a test failing due to multiple matches of / instead of just one. not sure why that changed in the dom?

OK, the culprit was:

https://github.com/hyparam/hyperparam-cli/pull/14/commits/79894fdd11d1c0d5c8b413e0f95ae393a5745ce0#diff-6af28ba0149c36fb2e0a4050a592e6c070d37f04359a949ccb21b39fa05e5583R42

Fixed.

File: packages/components/src/lib/tableProvider.ts (hunk 17 above)
Comment (severo) [reply packages/components/src/lib/tableProvider.ts]: OK. I changed the comments.

File: .github/workflows/ci.yml (hunk 14 above)
Comment (severo) [reply .github/workflows/ci.yml]: OK, I added it again. The issue was that... I had to build the dependency first before being able to run tsc.

File: packages/components/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (severo) [reply packages/components/README.md]: done.

**Comment by severo** [comment]
Merging! Tomorrow, I'll start creating PRs (or issues) for the points listed in the description
